### ISSUE TITLE
Uninstall target removes all CMake config files and installation folder

### ISSUE THREAD

**Issue by claudiofantacci** [PointCloudLibrary/pcl#2615]
## Your Environment
* Operating System and version: Windows 10
* Compiler: MSVC
* PCL Version: 1.9.0-dev

## Context
Issuing `ninja uninstall` the CMake config installation files and folder are completely removed, including config files and folder of other libraries.
This is the very last uninstall log row of the uninstall target:
```
Uninstalling "C:/Users/cfantacci/code-install/cmake"
```

## Expected Behavior
Remove only PCL-related config files and folders.

## Current Behavior
It broke my setup because it removed che CMake config folder.

## Code to Reproduce
Just issue `ninja uninstall`, but please don't. 😃 

## Possible Solution
Under investigation.

**Comment by SergioRAgostinho** [comment]
On linux the solution would be to issue a `rmdir` for folders so it erases the dir only if empty.  The code for the uninstallation section is here

https://github.com/PointCloudLibrary/pcl/blob/d0b92fa48d495724b155e2544272c48d6d9514ff/CMakeLists.txt#L229-L233

**Comment by claudiofantacci** [comment]
And this is the incriminated code
https://github.com/PointCloudLibrary/pcl/blob/d0b92fa48d495724b155e2544272c48d6d9514ff/cmake/uninstall_target.cmake.in#L35-L49

The [`remove_directory`](https://cmake.org/cmake/help/latest/manual/cmake.1.html#command-line-tool-mode) deletes a folder irrespective of whether it is empty or not.

**Comment by claudiofantacci** [comment]
In [YARP](https://github.com/robotology/yarp) we use [this useful CMake module](https://github.com/robotology/yarp/blob/master/extern/ycm/ycm-0.9.0.8/modules/AddUninstallTarget.cmake) to create a target deletes files only rather than folder, that is actually part/extracted from the [YCM](https://github.com/robotology/ycm) project that we also maintain. Everything is plance CMake, hence no extra dependency is needed. It may be of help for PCL also.

cc @drdanz

**Comment by traversaro** [comment]
Note that the YCM module mentioned by @claudiofantacci is actually based on an old version of the `uninstall` target source code described in CMake's FAQ, the updated version is available in https://gitlab.kitware.com/cmake/community/wikis/FAQ#can-i-do-make-uninstall-with-cmake . Both the old and new version just delete files and they do not delete directories.

Related YCM issue: https://github.com/robotology/ycm/issues/169 .

**Comment by claudiofantacci** [comment]
A possible change for PCL would be to strip those blocks that remove directories irrespective to their content. The one for the CMake config files is definitely wrong, or at least error prone with a standard installation. The other two blocks about include folders and documentation seems to be ok. I'm quite sure the block of the include folders are only for PCL created folders, while I nevere tested the one of the documentation.

**Comment by UnaNancyOwen** [comment]
You can uninstall only installed files based on install_manifest.txt. 
But, that list is not a complete at least on Windows. It is only listed files that current build configuration (either only debug or release). If it is a complete file, I think it will be possible.

It is the result of install with Release configuration, and after install with Debug configuration.It has been 
It is overwritten, only listed PCL library of Debug configuration.
https://gist.github.com/UnaNancyOwen/752668fd35553a58dbd5545fc6389e77#file-install_manifest-txt

**Comment by claudiofantacci** [comment]
@UnaNancyOwen you are right. I usually do compile and install Release and Debug from two different build folders, but the problem is present if your are compiling and installing from Visual Studio.
Having said this, the current uninstall target is still wrong and dangerous, at least for removing the config folder.

**Comment by claudiofantacci** [comment]
I can make a PR with one of the following solutions:
1. Replace https://github.com/PointCloudLibrary/pcl/blob/d0b92fa48d495724b155e2544272c48d6d9514ff/cmake/uninstall_target.cmake.in#L37-L49
with (just check the first 5 lines)
```cmake
file(GLOB_RECURSE CMAKE_CONFIG_FOLDER_FILES FOLLOW_SYMLINKS
     LIST_DIRECTORIES false
     "@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@/*")
list(LENGTH CMAKE_CONFIG_FOLDER_FILES CMAKE_CONFIG_FOLDER_FILES_NUMBER)
if (CMAKE_CONFIG_FOLDER_FILES_NUMBER GREATER 0)
    message(STATUS "Uninstalling \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\"")
    if(EXISTS "@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@")
        exec_program("@CMAKE_COMMAND@"
                     ARGS "-E remove_directory \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\""
                     OUTPUT_VARIABLE rm_out RETURN_VALUE rm_retval)
        if(NOT "${rm_retval}" STREQUAL 0)
            message(FATAL_ERROR "Problem when removing \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\"")
        endif()
    else(EXISTS "@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@")
        message(STATUS "Directory \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\" does not exist.")
    endif()
endif()
```
to check whether the `cmake` folder is empty or not. If it is empty, we also delete the folder.
2. Delete https://github.com/PointCloudLibrary/pcl/blob/d0b92fa48d495724b155e2544272c48d6d9514ff/cmake/uninstall_target.cmake.in#L37-L49 altogether.

I have no particular preference. My two cents is that deleting folder is _always_ potentially dangerous so I would avoid doing that. Consequently I find option 2 safer.

I also checked the remaining CMake code that delates PCL folders and they are safe, addressing @UnaNancyOwen concern about Windows platform.

**Comment by SergioRAgostinho** [comment]
I'm a little confused. I usually install PCL locally i.e. `pcl/install-1.8.0`. Something similar to the build folder. This install directory starts empty. After installing and uninstalling these are the remains I have
```
$ tree install-1.8.0/
install-1.8.0/
├── include
├── lib
│   └── pkgconfig
└── share
```
As you can see common folders like include, lib, share are preserved. Only pcl-major.minor seem to be deleted.

So this current issue is only observed in Windows correct?

Can any of you post the folder structure adopted for the windows installation rules please.

**Comment by claudiofantacci** [comment]
@SergioRAgostinho on Windows is different (spoiler: CMake config files are installed in a common `cmake` folder). I'll post the Windows tree structure tomorrow.

The problem can be traced back to `find_package` command that, until CMake version 3.6, used different paths to look for config files under Unix-based systems and Windows, see [here (scroll until middle page)](https://cmake.org/cmake/help/v3.6/command/find_package.html).
More modern CMake, from version 3.7+ can use new and same tree folder configuration for both Unix and Windows, see [here](https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure).

**Comment by traversaro** [comment]
Related CMake issue detailing why on CMake < 3.7 it was a good idea to install CMake config files in `<prefix>/cmake` on Windows: https://gitlab.kitware.com/cmake/cmake/issues/16212 .

**Comment by claudiofantacci** [comment]
This is the tree of my installation folder under Windows
```
> ls .\code-install\

    Directory: C:\Users\cfantacci\code-install

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        13-Nov-18     12:47                bin
d-----        13-Nov-18     12:47                cmake
d-----        14-Nov-18     12:24                include
d-----        13-Nov-18     12:47                lib
d-----        04-Jan-18     17:37                robotology
d-----        21-Aug-18     11:35                share
```
and all the configs are under the folder `cmake`, plainly copied (i.e. without folders with package names).

> Also `PCLCONFIG_INSTALL_DIR` doesn't appear in my CMake cache so I assume it can't be configured by the user to be empty. Basically we can't ever be in situation 1.

With situation 1 you mean this `<prefix>/    (W)`?

> I prefer to go with your first option

Ok, I'll submit a PR asap.

**Comment by SergioRAgostinho** [comment]
> > Also `PCLCONFIG_INSTALL_DIR` doesn't appear in my CMake cache so I assume it can't be configured by the user to be empty. Basically we can't ever be in situation 1.
> 
> With situation 1 you mean this `<prefix>/ (W)`?

Exactly :+1: 

> > I prefer to go with your first option
> 
> Ok, I'll submit a PR asap.

Don't commit to it just yet. Let me gather some more opinions on this. Ping  @PointCloudLibrary/maintainers  .

**Comment by claudiofantacci** [comment]
Sure, not a problem.

**Comment by taketwo** [comment]
I generally agree with the proposed solution. However in your snippet:

```cmake
file(GLOB_RECURSE CMAKE_CONFIG_FOLDER_FILES FOLLOW_SYMLINKS
     LIST_DIRECTORIES false
     "@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@/*")
list(LENGTH CMAKE_CONFIG_FOLDER_FILES CMAKE_CONFIG_FOLDER_FILES_NUMBER)
if (CMAKE_CONFIG_FOLDER_FILES_NUMBER GREATER 0)
    message(STATUS "Uninstalling \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\"")
    if(EXISTS "@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@")
        exec_program("@CMAKE_COMMAND@"
                     ARGS "-E remove_directory \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\""
                     OUTPUT_VARIABLE rm_out RETURN_VALUE rm_retval)
        if(NOT "${rm_retval}" STREQUAL 0)
            message(FATAL_ERROR "Problem when removing \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\"")
        endif()
    else(EXISTS "@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@")
        message(STATUS "Directory \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\" does not exist.")
    endif()
endif()
```

You first check if the directory is not empty, and then in a nested `if` you check if the directory exists at all, does not make sense.

**Comment by claudiofantacci** [comment]
@taketwo you are right and there is also another error, `GRATER`  should be `EQUAL`.
That was just an idea to keep the feature of deleting the CMake installed config folder.
Sorry for making confusion.

This is how I've implemented the new part of the uninstall target and now testing it.
```cmake
# remove pcl directory in share (removes all files in it!)
# created by CMakeLists.txt for PCLConfig.cmake
if(EXISTS "@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@")
  file(GLOB_RECURSE CMAKE_CONFIG_FOLDER_FILES FOLLOW_SYMLINKS
       LIST_DIRECTORIES false
       "@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@/*")
  list(LENGTH CMAKE_CONFIG_FOLDER_FILES CMAKE_CONFIG_FOLDER_FILES_NUMBER)
  if(CMAKE_CONFIG_FOLDER_FILES_NUMBER EQUAL 0)
    message(STATUS "Uninstalling \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\"")
    exec_program("@CMAKE_COMMAND@"
                 ARGS "-E remove_directory \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\""
                 OUTPUT_VARIABLE rm_out RETURN_VALUE rm_retval)
    if(NOT "${rm_retval}" STREQUAL 0)
      message(FATAL_ERROR "Problem when removing \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\"")
    endif()
  endif()
else(EXISTS "@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@")
  message(STATUS "Directory \"@CMAKE_INSTALL_PREFIX@/@PCLCONFIG_INSTALL_DIR@\" does not exist.")
endif()
```

**Comment by taketwo** [comment]
LGTM, thanks!
